### DISCUSSION TITLE
Alternative to ApskRadioMedium

### DISCUSSION THREAD

**Discussion by n-jay** [inet-framework/inet#976] [General]
Hi all,
I'm working on simulating a sensor network based on the IEEE 802.15.4 standard and having trouble getting nodes to _receive packets_. To elaborate further, when I set a debug point within the `processMessage` method of my routing protocol, it is **not** getting activated.

The transmission media parameters of my _ini_ file are as follows:

```
*.radioMedium.typename = "Ieee802154NarrowbandScalarRadioMedium"
*.radioMedium.backgroundNoise.power = -100dBm
*.radioMedium.mediumLimitCache.centerFrequency = 2GHz

**.constraintAreaMinX = 1m				
**.constraintAreaMaxX = 100m				
**.constraintAreaMinY = 1m				
**.constraintAreaMaxY = 100m				

*.host*.wlan[0].typename = "Ieee802154NarrowbandInterface"
*.host*.wlan[0].radio.typename = "ApskScalarRadio"
*.host*.wlan[0].radio.centerFrequency = 2.5GHz				
*.host*.wlan[0].radio.bandwidth = 2MHz
*.host*.wlan[0].radio.transmitter.power = 60.5mW 			
*.host*.wlan[0].radio.transmitter.preambleDuration = 192us	
*.host*.wlan[0].radio.transmitter.headerLength = 6B			
*.host*.wlan[0].radio.receiver.sensitivity = -98dBm 
*.host*.wlan[0].radio.receiver.energyDetection = -85dBm
*.host*.wlan[0].radio.receiver.snirThreshold = 4dB 
*.host*.wlan[0].radio.displayCommunicationRange = true


*.baseStation.wlan[0].typename = "Ieee802154NarrowbandInterface"
*.baseStation.wlan[0].radio.typename = "ApskScalarRadio"
*.baseStation.wlan[0].radio.centerFrequency = 2.5GHz				
*.baseStation.wlan[0].radio.bandwidth = 2MHz
*.baseStation.wlan[0].radio.transmitter.power = 60.5mW		
*.baseStation.wlan[0].radio.transmitter.preambleDuration = 192us	
*.baseStation.wlan[0].radio.transmitter.headerLength = 6B
*.baseStation.wlan[0].radio.receiver.sensitivity = -98dBm
*.baseStation.wlan[0].radio.receiver.energyDetection = -85dBm
*.baseStation.wlan[0].radio.receiver.snirThreshold = 4dB
*.baseStation.wlan[0].radio.displayCommunicationRange = true

*.host*.**.bitrate = 250kbps								
*.baseStation.**.bitrate = 250kbps							
```
I have not set any pathloss algorithms.

For more context, the submodules in my NED file are as follows:

```
    submodules:
        configurator: Ipv4NetworkConfigurator {
            @display("p=100,100");
        }
        radioMedium: <default("UnitDiskRadioMedium")> like IRadioMedium {
            @display("p=100,100");
        }
        physicalEnvironment: PhysicalEnvironment {
            @display("p=100,100");
        }
        host[numNodes]: <default("WirelessHost")> like INetworkNode {
            @display("i=misc/sensor2");
        }
        baseStation: <default("WirelessHost")> like INetworkNode {
            @display("p=50,50;i=misc/sensorgateway");
        }
```

I'm quite certain this issue is not stemming from my routing protocol code as the exact same code ran successfully in INET 4.2.5 (on OMNET v5) but with the difference of the radio medium in the ini file being `ApskScalarRadioMedium`. Since this module seems to have been deprecated in the latest releases of INET I opted to use the `Ieee802154NarrowbandScalarRadioMedium` as shown above assuming it should work fine.
Since this is the only difference, I'm fairly confident it's the cause of this issue. 

Would really appreciate some thoughts on this, perhaps an alternative to the earlier `ApskScalarRadioMedium` module that fits my requirement.

**Comment by levy** [comment]
You should try replacing the old ApskScalarRadioMedium/ApskScalarRadio with RadioMedium/ApskRadio. The signalAnalogRepresentation parameter controls how the signals are represnted in the analog domain. It is set to scalar by default, so the default should work just fine.

**Reply by n-jay** [reply]
@levy I changed the `.radioMedium.typename` to `RadioMedium`, and `.wlan[0].radio.typename` to `ApskRadio` but it still doesn't hit the debug point. 

Should the `.wlan[0].typename` be kept as `Ieee802154NarrowbandInterface` as in the code above, or should that also be changed? The use of the `Ieee802154NarrowbandInterface` is actually carried forward from my earlier simulation, accommodating the `ApskScalarRadioMedium/ApskScalarRadio` configuration.

**Comment by levy** [comment]
It should be the one that you used before you switched to Ieee802154NarrowbandInterface. Basically, I'm saying that the scalar APSK radio is still supported. Those NED types are removed because the scalar/dimenstional thing became a parameter instead of being hardcoded in the NED type. You can see some examples under examples/wireless/nic/omnetpp.ini

**Reply by n-jay** [reply]
> It should be the one that you used before you switched to Ieee802154NarrowbandInterface.

I think there's a misunderstanding, I _didn't_ switch to `Ieee802154NarrowbandInterface`, 
I _continued_ using it from my earlier OMNET v5 simulation with `ApskRadioMedium` as the `.radioMedium.typename` and `.wlan[0].radio.typename` as `ApskScalarRadio`. 
Basically my original simulation ini looked like this:

-  `*.radioMedium.typename = "ApskRadioMedium"`
-  `.wlan[0].typename = "Ieee802154NarrowbandInterface"`
-  `.wlan[0].radio.typename = "ApskScalarRadio"`

I switched `ApskRadioMedium` to `Ieee802154NarrowbandScalarRadioMedium` because `ApskRadioMedium` seems to be deprecated.

> Those NED types are removed because the scalar/dimenstional thing became a parameter instead of being hardcoded in the NED type. You can see some examples under examples/wireless/nic/omnetpp.ini

Noted. I'll check the examples directory and try to replicate that. Thanks!
Seems like the **AbstractApskScalarRadio** config seems to match my requirement. Will update this thread on how it goes.

**Comment by n-jay** [comment]
**RESOLVED**
Turns out the issue was **NOT** in the simulation parameters, but rather using a some obsolete code in the `initialize()` method within the protocol:
```
} else if (stage == INITSTAGE_ROUTING_PROTOCOLS) {
       registerService(Protocol::manet, nullptr, gate("ipIn"));
       registerProtocol(Protocol::manet, gate("ipOut"), nullptr);
}
```

The corrected form is:
```
} else if (stage == INITSTAGE_ROUTING_PROTOCOLS) {
       registerProtocol(Protocol::manet, gate("ipOut"), gate("ipIn"));
}
```
